compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -200,4 +210,52 @@ impl<'tcx> InstCombineContext<'tcx, '_> {
         });
         terminator.kind = TerminatorKind::Goto { target: destination_block };
     }
+
+    fn combine_intrinsic_assert(
+        &self,
+        terminator: &mut Terminator<'tcx>,
+        _statements: &mut Vec<Statement<'tcx>>,
+    ) {
+        let TerminatorKind::Call { func, target: Some(destination_block), .. } = &mut terminator.kind  else { return; };
+        let func_ty = func.ty(self.local_decls, self.tcx);
+        let Some((intrinsic_name, substs)) = resolve_rust_intrinsic(self.tcx, func_ty) else {
+            return;
+        };
+        if substs.is_empty() {
+            return;
+        }
+        let ty = substs.type_at(0);
+        let Ok(layout) = self.tcx.layout_of(self.param_env.and(ty)) else { return; };
```

Comment:
this is done on every intrinsic with at least one generic parameter, should probably only do it on the three ones below. Maybe do an early check for the intrinsic kind and make that return a `fn(TyCtxt<'tcx>, TyLayout<'tcx>> -> bool` function pointer for these three intrinsics, and then call that one to check whether to do the terminator replacing.